widget/widget.go

```diff
@@ -13,51 +15,95 @@ type BaseWidget struct {
 	position fyne.Position
 	Hidden   bool
 
-	impl fyne.Widget
+	impl         fyne.Widget
+	propertyLock sync.RWMutex
 }
 
 // ExtendBaseWidget is used by an extending widget to make use of BaseWidget functionality.
 func (w *BaseWidget) ExtendBaseWidget(wid fyne.Widget) {
+	w.propertyLock.Lock()
+	defer w.propertyLock.Unlock()
+
 	if w.impl != nil {
 		return
 	}
 
 	w.impl = wid
 }
 
+// ReadFields provides a guaranteed thread safe way to access widget fields.
+func (w *BaseWidget) ReadFields(f func()) {
+	w.propertyLock.RLock()
+	defer w.propertyLock.RUnlock()
+
+	f()
+}
+
+// SetFieldsAndRefresh helps to make changes to a widget that should be followed by a refresh.
+// This method is a guaranteed thread-safe way of directly manipulating widget fields.
+func (w *BaseWidget) SetFieldsAndRefresh(f func()) {
+	w.propertyLock.Lock()
+	f()
+	w.propertyLock.Unlock()
+
+	w.Refresh()
+}
+
 // Size gets the current size of this widget.
 func (w *BaseWidget) Size() fyne.Size {
+	w.propertyLock.RLock()
+	defer w.propertyLock.RUnlock()
+
 	return w.size
 }
 
 // Resize sets a new size for a widget.
 // Note this should not be used if the widget is being managed by a Layout within a Container.
 func (w *BaseWidget) Resize(size fyne.Size) {
-	if w.size == size {
+	var baseSize fyne.Size
+	w.ReadFields(func() {
+		baseSize = w.size
+	})
+	if baseSize == size {
 		return
 	}
-	w.size = size
 
-	if w.impl == nil {
+	var impl fyne.Widget
+	w.ReadFields(func() {
+		w.size = size
+
+		impl = w.impl
+	})
+	if impl == nil {
 		return
 	}
-	cache.Renderer(w.impl).Layout(size)
+	cache.Renderer(impl).Layout(size)
 }
 
 // Position gets the current position of this widget, relative to its parent.
 func (w *BaseWidget) Position() fyne.Position {
+	w.propertyLock.RLock()
+	defer w.propertyLock.RUnlock()
```

Comment:
Use the helper methods to avoid lock/unlock

Reply:
As above